Integrate with AWS Lambda - Requirements file, .gitignore updates, integrate SSM, SAM template.

javydekoning opened the pull request:
adding requirements file and updating .gitignore

wesselt commented on bunq2ynab.py:
```diff
@@ -1,11 +1,19 @@
 from lib.sync import Sync
 from lib.config import config
+import os
```
There is a new file aws_lambda.py that was meant for the AWS handler function. Why not use that instead of giving bunq2ynab.py two jobs?

javydekoning replied:
Code portability. In this case we only need to maintain a single application entry point. We can split it up if you prefer, but this looks the easiest to me.

wesselt commented on lib/config.py:
```diff
@@ -4,6 +4,7 @@
 import os
 import sys
 
+from lib.parameter_store import ParameterStore
```
Aha, so ParameterStore gets the "environment" of the lambda function. Looks good

javydekoning replied:
Correct, see `template.yml` (used for deployment) this sets the environment variable [1] and creates the config in SSM. [2]

1. https://github.com/wesselt/bunq2ynab/pull/22/commits/38607aefd4b9246ce23fe271f2aa062c3dc3ac9a#diff-363d481fe07d1094db2338998d381b71R46-R47
2. https://github.com/wesselt/bunq2ynab/pull/22/commits/38607aefd4b9246ce23fe271f2aa062c3dc3ac9a#diff-363d481fe07d1094db2338998d381b71R64-R80

wesselt commented on requirements.txt:
```diff
@@ -0,0 +1,3 @@
+pyOpenSSL
+requests
+boto3
```
The "miniupnpc" library is an optional requirement too, used when you're behind a firewall (like most Raspberry Pis')

Which tool/app uses requirements.txt ?

javydekoning replied:
"Requirements files" are files containing a list of items to be installed using pip install like `pip install -r requirements.txt`

Refer to [the documentation here](https://pip.pypa.io/en/stable/user_guide/#requirements-files). This makes setup easier for users. We should also version pin, which I'll add later.

`sam build` also uses this file, it will package all the dependencies required by the Lambda function.

wesselt replied:
Cool, the requirements look useful, I see a lot of other Python projects have a requirements file.

So in AWS the config.json file is essentially in the ParameterStore. I was wondering about state.json, which basically stores the bunq installation. That's required because a bunq api token is only valid for 4 hours. Where should the lambda store its state?

javydekoning replied:
It stored on the scratch space until the function is 'recycled' which happens frequently. When that happens bunq2ynab starts 'fresh' and the `state.json` is recreated.

wesselt replied:
> When that happens bunq2ynab starts 'fresh' and the `state.json` is recreated.

I think you can only create new installations for 4 hours after you create the API token. The installation is basically the private key + installation token that's stored in state.json. So I'd be surprised if that worked for a long time.

It's only the session token that can be endlessly refreshed.

javydekoning replied:
If that is true, we would need to store those in SSM as well. That is how I'm doing it today and would not be a problem.

However, sometimes the install is no-longer valid and we somehow need to detect and refresh the state json. Could you point me to the code where you detect and handle that?

wesselt replied:
I've never seen an installation token become invalid just because it's old. I think you have only 4 hours after creating an API key to create an installation token.

As far as I know the only thing that becomes invalid over time is the session token. That's detected at https://github.com/wesselt/bunq2ynab/blob/config/lib/bunq.py#L189.  The code will request a new session token and ask the State class to store it.

wesselt commented:
Looks good 👍  Added some comments!  I'll try to set up a lambda at AWS to see it in action

javydekoning commented:
> Looks good 👍 Added some comments! I'll try to set up a lambda at AWS to see it in action

Cool, let me know if you need any help with that!

wesselt commented:
> Cool, let me know if you need any help with that!

Got to this point.  It's probably Google-able, will try further this weekend!

2020-08-13 14:50:53,368 \| CRITICAL \| config.py:76 \| read_ssm_config \| Error loading configuration from SSM Parameter: bunq2ynab-config: An error occurred (AccessDeniedException) when calling the GetParameter operation: User: arn:aws:sts::122776901776:assumed-role/bunq2ynab-1aws-2aws-role-254ipodc/bunq2ynab-1aws-2aws is not authorized to perform: ssm:GetParameter on resource: arn:aws:ssm:eu-central-1:122776901776:parameter/bunq2ynab-config

javydekoning commented:
Looks like you didn't deploy from the template. Your lambda is missing a role with permission to read from ssm. Let me send you instructions tomorrow :-)

wesselt commented:
Yeah, I thought if I'd take the template from the serverless app store, I'd get the old version?

Looking forward to how to add the permissions.

javydekoning commented:
> Yeah, I thought if I'd take the template from the serverless app store, I'd get the old version?
> 
> 
> 
> Looking forward to how to add the permissions.

Yes, I didn't publish this version yet as I didn't have time to run tests. I'll reach out tomorrow morning.

javydekoning commented:
To build and deploy you need [SAM](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-install.html). (Users will not need this, they will be able to install from the repo).

Instructions to get it running:
 
```
sam build
sam deploy --guided

Stack Name [sam-app]: b2ytest
AWS Region [us-east-1]: eu-west-1
Parameter BunqApiToken []: <removed>
Parameter YnabAccessToken []: <removed>
Confirm changes before deploy [y/N]: y
#SAM needs permission to be able to create roles to connect to the resources in your template
Allow SAM CLI IAM role creation [Y/n]: y
Save arguments to samconfig.toml [Y/n]: n
```

- Now in your [CloudFormation Console](https://eu-west-1.console.aws.amazon.com/cloudformation/) you'll find the stack being deployed.
- Go to [Systems Manager Parameter Store](https://eu-west-1.console.aws.amazon.com/systems-manager/parameters/)
- Edit your configuration if needed.
- Now go and run your Lambda

It will immediately error out due to an argparse error, but I don't have time to debug it today. Maybe I can spend some time on it on Sunday. Shouldn't be to difficult.

wesselt commented:
Just merged this request from the command line. Because I rebased the pull request doesn't seem to notice?  Or now it does.

I'm still trying to get SAM to install. I've got it in Windows but then it doesn't work in WSL. Arch Linux doens't seem to have a package for it, and it doesn't work on my Pi.

I was able to manually add the rights to the parameters in IAM > Access management > Roles. Still have to figure out how this works with encrypted parameters.

Split SSM_PARAM into SSM_CONFIG_PARAM and SSM_STATE_PARAM. The latest version gets to a point where bunq says the API key is invalid. Curious how far you get!

wesselt commented:
Got some more tries in, it now works for me!  The "import boto3" is now in a separate function that is only called when SSM_CONFIG_PARAM or SSM_STATE_PARAM is set.

I changed it so it doesn't log the state or configuration and the keys and tokens in them.

javydekoning commented:
That's great! Thanks. I'm still a little busy this week (holiday coming up :-) ). When I have some room I'll migrate my sync over to this version and check whether all will be ok.

wesselt commented:
That'd be nice, checking is very useful!  Next up for me is a look at API gateways. Maybe we could add that to the template at some point.

javydekoning commented:
I checked and fixed a few things. 

- Added the correct permissions to template.json (for accessing the state in SSM).
- Moved config inside of handler function (If you don't a change in the config is not read until a new Lambda is created by AWS)
- Renamed the environment variables in `template.yaml` to match the code.

I'll now have it run for a couple of days to make sure everything works as expected. Adding a callback would be super easy.

wesselt commented:
Thanks for testing!  I can't see your code changes. Maybe because this pull request is already merged?

javydekoning commented:
I comitted to config branch: https://github.com/wesselt/bunq2ynab/commit/08a464654074631e22b94feaa0869bbd329325a2

javydekoning commented:
Update, running fine until now. No issues. If this continues to work I'll push this version to the Serverless Application Repository. I'll create issues for the remaining tasks.
